### ISSUE TITLE
Generic add to F-Droid request

### ISSUE THREAD

**Issue by ghost** [bisq-network/bisq-mobile#135]

**Comment by rodvar** [comment]
@endingisnight hey thanks for creating an issue proposing this! We are working hard on development of the MVP at the moment and adding to F-Droid was not part of the original effort estimation that the DAO approved, but obviously our apps are perfect candidates for it. We've added it to the next immediate milestone. Thanks 🙏

**Comment by rodvar** [comment]
@endingisnight hope you are doing great. we might start working on this earlier than expected, could you recommend the best way to do it? 

I found an [old blogpost](https://kaeruct.github.io/posts/2021/08/03/publishing-an-app-on-f-droid/) of a successful experience, would that be a good tutorial to follow?

Thanks in advance! 🙏

**Comment by rodvar** [comment]
So there are 2 ways of doing this, one is to host your own F-Droid server, with your apk and metadata 

## F-Droid self hosted

### Requirements

```
You have a good understanding of what [Free Software](https://www.gnu.org/philosophy/free-sw.html) means, and what F-Droid is for.
You already read and understand the [Inclusion Policy](https://f-droid.org/docs/Inclusion_Policy).
You already read and understand the [Repository Style Guide](https://f-droid.org/docs/Repository_Style_Guide).
You already read and understand [the relevant parts of the F-Droid documentation](https://f-droid.org/docs/Build_Metadata_Reference).
You know how to use [Git VCS](https://git-scm.com/), and know how a merge request (a.k.a. “pull request” in GitHub terminology) works in general.
You have an account on [GitLab](https://gitlab.com/) (NOTE: F-Droid CI runners are under Gitlab’s FOSS program, so there’s no need for you to pay for any CI time. If Gitlab starts asking for phone numbers or credit cards don’t submit anything, just leave a note in the MR so we know we need to trigger the CI.)
You have a local instance of the F-Droid server software, and you know what you are doing.
```

### How

 - Please refer to this [official guide](https://f-droid.org/docs/Setup_an_F-Droid_App_Repo/) , basically once you have the f-droid server on you need to lodge a [metadata merge request into their gitlab repo](https://gitlab.com/fdroid/fdroiddata/-/tree/master/metadata?ref_type=heads)

## F-Droid main public repo

### Requirements

 - apps need to be fully buildable from source (which we comply with already)
 - F-Droid team uses their own signature for the apk
 - they use apk not aab
 - a wait of ~2-3 weeks for review is anticipated.

### How

 - Submit an app inclusion request here -> https://gitlab.com/fdroid/rfp/-/issues

## Comparison:

| **feature/submission** | **Self-hosted** | **f-droid hosted** | 
|-|-|-| 
| Signing | our signature | theirs, however for some privacy intensive us like ours there are exceptions. |
| Review | same | same |
| Effort | some effort expected, we could use our test server for this purpose | minor, since our apps already builds from source |
| Timeline | once the server is up and running and the metadata is compliant, review happens much faster | addition queue takes at the very least 2 weeks |

**Comment by rodvar** [comment]
based on the above, might make sense to try to host a test apk on our github pages repo and test doing a review with it to see if the listing gets approved. In that way we use the easiest less costly path without compromising our signature.

@endingisnight do you think this is a winner approach for bisq apps? thanks!

**Comment by endingisnight** [comment]
@rodvar personally, I think that having it in the official repo is a good idea since (as a correction to your comparison table) you can use your own signature with [reproducible builds](https://f-droid.org/docs/Reproducible_Builds/) (RB). Though the build cycles (timeline) in the official repo is usually a weekly thing (though they can do emergency updates) timed with [TWIF](https://f-droid.org/en/2025/04/10/twif.html).

But if you're really worried about the update speed you can also add the app to @izzysoft 's [repo](https://apt.izzysoft.de/fdroid/index/info) which has daily updates and dosen't require the apps to be built from source. But I won't stop you from making your own repo.

Also a few worries about this repo:
1. there's no LICENSE file?
2. add [fastlane](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) for epic metadata
3. if you're gonna want RB (or for izzy) then they'll need your builds, signed by you probably in the releases of the repo (this also includes tagging your release commits)
4. Also you'll probably want to read [this part](https://android.izzysoft.de/articles/named/iod-scan-apkchecks#blobs) and remove the dependency info block.

**Comment by IzzySoft** [comment]
> which has daily updates and dosen't require the apps to be built from source

What else to build them from? 😜 But yes, we always distribute the APK provided by the resp. developers. And we also support reproducible builds, see e.g. [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients) – which run on a separate track and thus do not delay releases here. Also see the [IzzyOnDroid App Inclusion Policy](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Inclusion-Policy) for additional details.

But we'd need the APK provided at [Releases](https://github.com/bisq-network/bisq-mobile/releases) – ideally built from the commit the release tag points to (see: [hints for developers on reproducible builds](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Reproducible-Builds/RB-Hints-for-Developers)). And we'd need a license specified (searching the code: this would be AGPL-3.0-only, right?).

**Comment by rodvar** [comment]
hey @IzzySoft , @endingisnight - I think I've covered everything now.

I generated a `pre-release test` [here](https://github.com/bisq-network/bisq-mobile/releases/tag/aconnect_0.0.30) for one of our Android apps (`Android Bisq Connect v0.0.30`).

Just got a couple of questions for you:

1. first and foremost, are things looking compliant for FDroid now (added license & fastlane metadata pulled from play store. Is still not 100% the one we will use I think but close to)?
2. Our release builds are non-deterministic unless I disable `minimization` and `resource shrinking` which is obviously not great. Any tips on how to make it deterministic without disabling it? Or otherwise, how to keep our release signatures for the FDroid releases apks?

Thanks for the help again!

**Comment by IzzySoft** [comment]
First: thanks for providing the APK! I've sent it through our scanners, and it looks absolutely fine – on a first glance, definitely no show-stoppers for either IzzyOnDroid or F-Droid.

Next, I've tried to build the app. Failed: according to the embedded commit hash, the APK was built from commit 28733b18fe90f84c576de2cc037b4bad3227894d – but that commit does not exist here. So did you forget to push it? The APK should have been built from the commit the release tag points to, by the way. OK, so I've tried to build from the tagged commit – but that fails as well:

```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache state could not be cached: field `__inputFiles__` of task `:androidNode:extractIncludeReleaseProto` of type `com.google.protobuf.gradle.ProtobufExtract`: error writing value of type 'org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection'
> Could not resolve all files for configuration ':androidNode:releaseCompileClasspath'.
   > Could not find bisq:common:2.1.7.
     Required by:
         project :androidNode
   > Could not find bisq:i18n:2.1.7.
     Required by:
         project :androidNode
   > Could not find bisq:persistence:2.1.7.
...
```

Are really all sources needed to compile the app checked in and pushed to Github? If not, that would be a show-stopper for F-Droid. 

OK, seems to be my fault:  the readme says it needs bisq2 – but the instructions are incomplete (e.g. at the bisq2 repo it says "For Windows environments: replace ./gradlew with gradle.bat as the previous example shows" – but there is no single mention of gradlew at all. It just mentions how to **run** the desktop variants. Also, it's not clear which commit to check out from there – something that's important for [Reproducible Builds](https://reproducible-builds.org/) (for some background, please see e.g. [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Without a tag or commit specified, results depend on what time the build is done, as `HEAD` might have moved forward there.

Now, we could probably use `release.sh` – but that 1) doesn't seem to take care for bisq2, 2) wants a parameter passed (which, and in which order?), and 3) would produce an AAB – while we'd need an APK.


> added license & fastlane metadata

License looks fine! But where did the metadata go? Ah, found it: F-Droid will be unable to find it there, for them, `metadata` must be moved one level up (to the project root). Also see the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) for more details on this.

> Our release builds are non-deterministic unless I disable minimization and resource shrinking which is obviously not great. Any tips on how to make it deterministic without disabling it? Or otherwise, how to keep our release signatures for the FDroid releases apks?

Without reaching reproducible builds, you won't get your APKs to F-Droid with your own signature – F-Droid would create a signing key for your app and sign it with that. So no cross-updates then. Not sure why your builds should be non-deterministic as long as minimize & resource shrinking are enabled – we have hundreds of RB apps which have that on. Oh, and at this very moment, the next build of your app finished, let me look… oof, it was running into a timeout, build took longer than our default timeout. But at least it seems we'd be able to build:

```
> Task :androidNode:minifyReleaseWithR8

WARNING: R8: The method `void org.apache.commons.logging.impl.Log4JLogger.<clinit>()` does not type check and will be assumed to be unreachable.

Session terminated, killing shell... ...killed.

...

Buildtime: 1244 seconds (20:44 min)
```

Default timeout is 20 min. Seeing it seems to build (after all, it already reached R8 optimization), I've increased the timeout for your app to 30 min, and started the build again. Let's see where that goes. As I know my luck, it might succeed to build at about 28..29 minutes, and I guessed the resulting file name wrong, so the APK will simply be destroyed together with the (ephemeral) container it is build in… This recipe so far:

```yaml
build:
  - echo git reset --soft 28733b18fe90f84c576de2cc037b4bad3227894d
  - touch local.properties
  - sed -r '/signingConfigs.getByName/d' -i androidClient/build.gradle.kts
  - git clone --recurse-submodules https://github.com/bisq-network/bisq2.git
  - pushd bisq2
  - ./gradlew publishAll
  - popd
  - ./gradlew assembleRelease
```

using OpenJDK 17 on Debian bookworm. Oh, this time it completed after 15 minutes, but failed again.

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':androidNode:packageRelease'.
> A failure occurred while executing com.android.build.gradle.tasks.PackageAndroidArtifact$IncrementalSplitterRunnable
   > SigningConfig "release" is missing required property "storeFile".
```

Yuck, another `build.gradle` enforcing signing… wish developers would always wrap that inside `if (localProperties["KEYSTORE_PATH"] != null)` conditions. Makes no sense to declare `signingConfigs.getByName`if there's no storeFile, or do I miss something there? 😉 You've properly wrapped that in `signingConfigs` – but somehow, leaving the `getByName` in later, this does not suffice (not sure, maybe using `findByName` instead in the `buildType` section would take care for that, too – I vaguely remember having seen that once). OK, another run then… Yes! Build succeeded. And my "know my luck" was on the spot. Last run now, with the proper resulting "outfile" name… `androidClient/build/outputs/apk/release/Bisq Connect-0.0.30.apk` – with a space in the file name, for some "extra fun" in scripts 🙈 

> Our release builds are non-deterministic unless I disable minimization and resource shrinking

Haha, nope. Here's your diff:

```
  -rw-r--r--  0.0 unx       56 b-       52 defN 1981-01-01 01:01:02 aa2cdc4d META-INF/com/android/build/gradle/app-metadata.properties
- -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 e4c975b8 META-INF/version-control-info.textproto
+ -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 0ae90e91 META-INF/version-control-info.textproto
  -rw-r--r--  0.0 unx     7375 b-     7375 stor 1981-01-01 01:01:02 7c699c3e assets/dexopt/baseline.prof
```

It's only that missing commit (probably you've built, and after the build succeeded, squashed and tagged). In other words: here's your recipe for a reproducible build if the commit hash would match:

```yaml
build:
  - echo git reset --soft 28733b18fe90f84c576de2cc037b4bad3227894d
  - touch local.properties
  - sed -r '/signingConfigs.getByName/d' -i androidClient/build.gradle.kts
  - sed -r '/signingConfigs.getByName/d' -i androidNode/build.gradle.kts
  - git clone --recurse-submodules https://github.com/bisq-network/bisq2.git
  - pushd bisq2
  - ./gradlew publishAll
  - popd
  - ./gradlew assembleRelease
  - mv androidClient/build/outputs/apk/release/Bisq*Connect-0.0.30.apk /outputs/unsigned.apk
```

Buildtime: 932 seconds (15:32 min)

**Comment by rodvar** [comment]
> First: thanks for providing the APK! I've sent it through our scanners, and it looks absolutely fine – on a first glance, definitely no show-stoppers for either IzzyOnDroid or F-Droid.
> 
> Next, I've tried to build the app. Failed: according to the embedded commit hash, the APK was built from commit [28733b1](https://github.com/bisq-network/bisq-mobile/commit/28733b18fe90f84c576de2cc037b4bad3227894d) – but that commit does not exist here. So did you forget to push it? The APK should have been built from the commit the release tag points to, by the way. OK, so I've tried to build from the tagged commit – but that fails as well:
> 
> ```
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Configuration cache state could not be cached: field `__inputFiles__` of task `:androidNode:extractIncludeReleaseProto` of type `com.google.protobuf.gradle.ProtobufExtract`: error writing value of type 'org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection'
> > Could not resolve all files for configuration ':androidNode:releaseCompileClasspath'.
>    > Could not find bisq:common:2.1.7.
>      Required by:
>          project :androidNode
>    > Could not find bisq:i18n:2.1.7.
>      Required by:
>          project :androidNode
>    > Could not find bisq:persistence:2.1.7.
> ...
> ```
> 
> Are really all sources needed to compile the app checked in and pushed to Github? If not, that would be a show-stopper for F-Droid.
> 
> OK, seems to be my fault: the readme says it needs bisq2 – but the instructions are incomplete (e.g. at the bisq2 repo it says "For Windows environments: replace ./gradlew with gradle.bat as the previous example shows" – but there is no single mention of gradlew at all. It just mentions how to **run** the desktop variants. Also, it's not clear which commit to check out from there – something that's important for [Reproducible Builds](https://reproducible-builds.org/) (for some background, please see e.g. [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Without a tag or commit specified, results depend on what time the build is done, as `HEAD` might have moved forward there.
> 
> Now, we could probably use `release.sh` – but that 1) doesn't seem to take care for bisq2, 2) wants a parameter passed (which, and in which order?), and 3) would produce an AAB – while we'd need an APK.
> 
> > added license & fastlane metadata
> 
> License looks fine! But where did the metadata go? Ah, found it: F-Droid will be unable to find it there, for them, `metadata` must be moved one level up (to the project root). Also see the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane) for more details on this.
> 
> > Our release builds are non-deterministic unless I disable minimization and resource shrinking which is obviously not great. Any tips on how to make it deterministic without disabling it? Or otherwise, how to keep our release signatures for the FDroid releases apks?
> 
> Without reaching reproducible builds, you won't get your APKs to F-Droid with your own signature – F-Droid would create a signing key for your app and sign it with that. So no cross-updates then. Not sure why your builds should be non-deterministic as long as minimize & resource shrinking are enabled – we have hundreds of RB apps which have that on. Oh, and at this very moment, the next build of your app finished, let me look… oof, it was running into a timeout, build took longer than our default timeout. But at least it seems we'd be able to build:
> 
> ```
> > Task :androidNode:minifyReleaseWithR8
> 
> WARNING: R8: The method `void org.apache.commons.logging.impl.Log4JLogger.<clinit>()` does not type check and will be assumed to be unreachable.
> 
> Session terminated, killing shell... ...killed.
> 
> ...
> 
> Buildtime: 1244 seconds (20:44 min)
> ```
> 
> Default timeout is 20 min. Seeing it seems to build (after all, it already reached R8 optimization), I've increased the timeout for your app to 30 min, and started the build again. Let's see where that goes. As I know my luck, it might succeed to build at about 28..29 minutes, and I guessed the resulting file name wrong, so the APK will simply be destroyed together with the (ephemeral) container it is build in… This recipe so far:
> 
> build:
>   - echo git reset --soft 28733b18fe90f84c576de2cc037b4bad3227894d
>   - touch local.properties
>   - sed -r '/signingConfigs.getByName/d' -i androidClient/build.gradle.kts
>   - git clone --recurse-submodules https://github.com/bisq-network/bisq2.git
>   - pushd bisq2
>   - ./gradlew publishAll
>   - popd
>   - ./gradlew assembleRelease
> using OpenJDK 17 on Debian bookworm. Oh, this time it completed after 15 minutes, but failed again.
> 
> ```
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':androidNode:packageRelease'.
> > A failure occurred while executing com.android.build.gradle.tasks.PackageAndroidArtifact$IncrementalSplitterRunnable
>    > SigningConfig "release" is missing required property "storeFile".
> ```
> 
> Yuck, another `build.gradle` enforcing signing… wish developers would always wrap that inside `if (localProperties["KEYSTORE_PATH"] != null)` conditions. Makes no sense to declare `signingConfigs.getByName`if there's no storeFile, or do I miss something there? 😉 You've properly wrapped that in `signingConfigs` – but somehow, leaving the `getByName` in later, this does not suffice (not sure, maybe using `findByName` instead in the `buildType` section would take care for that, too – I vaguely remember having seen that once). OK, another run then… Yes! Build succeeded. And my "know my luck" was on the spot. Last run now, with the proper resulting "outfile" name… `androidClient/build/outputs/apk/release/Bisq Connect-0.0.30.apk` – with a space in the file name, for some "extra fun" in scripts 🙈
> 
> > Our release builds are non-deterministic unless I disable minimization and resource shrinking
> 
> Haha, nope. Here's your diff:
> 
> ```
>   -rw-r--r--  0.0 unx       56 b-       52 defN 1981-01-01 01:01:02 aa2cdc4d META-INF/com/android/build/gradle/app-metadata.properties
> - -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 e4c975b8 META-INF/version-control-info.textproto
> + -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 0ae90e91 META-INF/version-control-info.textproto
>   -rw-r--r--  0.0 unx     7375 b-     7375 stor 1981-01-01 01:01:02 7c699c3e assets/dexopt/baseline.prof
> ```
> 
> It's only that missing commit (probably you've built, and after the build succeeded, squashed and tagged). In other words: here's your recipe for a reproducible build if the commit hash would match:
> 
> build:
>   - echo git reset --soft 28733b18fe90f84c576de2cc037b4bad3227894d
>   - touch local.properties
>   - sed -r '/signingConfigs.getByName/d' -i androidClient/build.gradle.kts
>   - sed -r '/signingConfigs.getByName/d' -i androidNode/build.gradle.kts
>   - git clone --recurse-submodules https://github.com/bisq-network/bisq2.git
>   - pushd bisq2
>   - ./gradlew publishAll
>   - popd
>   - ./gradlew assembleRelease
>   - mv androidClient/build/outputs/apk/release/Bisq*Connect-0.0.30.apk /outputs/unsigned.apk
> Buildtime: 932 seconds (15:32 min)

hey @IzzySoft  - first of all, thanks so much for this detailed comment , it looks by the way its written that you basically logged your process reviewing 😄  haha it was fun to read!

I'm gonna re-read it in detail when I get more time, but on a first glance I'd like to comment:

 - happy to hear the generated apk is fine as is, so we can leave our release process as is with no changes then :) 
 - the testing apk I provided is only for the Bisq Connect Android app (`androidClient` folder). I started with that one cause, as you already got a feel for it, the androidNode is more complex to build as it depends on bisq2 J17 compatible jars generation.
 - therefore, the command you need to build for the androidClient (which I used to generate the uploaded apk) is `./gradlew androidClient:clean androidClient:assembleRelease` - we probably should add this in the README.md ...
 - One thing we need to address somehow, is that from your comments it looks like F-droid doesn't support multiple apps per project? Our codebase is a [KMP Compose](https://www.jetbrains.com/compose-multiplatform/) project, the kotlin sourcecode generates 3 apps in total. Each metadata therefore is linked to a specific folder (`androidClient` , `iosClient` and `androidNode` ). I will need to tap on your knowledge here how could we make it work?
 - sorry for the commit hash issue, I think what happend is that I linked the commit id whilst I was working on a release branch, and then merge that with a squash and merge which changed the commit id to `3839a62a27a67625ade4ed05292133228cd809f5` - will take that into account for the next one

Excited to hear your comments, thanks again!

**Comment by IzzySoft** [comment]
> it looks by the way its written that you basically logged your process reviewing

That's indeed how it was created – with a little cleanup afterwards (e.g. removing most of the errors from the first run, as they were not relevant: the error was mine). I thought doing it this way helps you see how another person would go based on what's there. First-hand experience: if I write up my docs I often take things granted as I know how they work, if I wrote the software.  So this could offer a possibility to see it with an "independent pair of eyes" 😉

> happy to hear the generated apk is fine as is, so we can leave our release process as is with no changes then

Looks like, yes.

> the testing apk I provided …

The process also produced the other APK, but I ignored it for this case. It was more or less a POC, showing the core points (like, the APK should have been built from the tagged commit, the commit needs to be present, it indeed seems to be deterministic, where some instructions would be helpful, the successful recipe to forward).

> F-droid doesn't support multiple apps per project?

Which platform does? It would be separate apps. Same on PlayStore, no? If multiple apps are needed, the descriptions need to be clear on that fact. And at least currently, F-Droid is about Android apps, so it would not cover your iosClient at all.

> and then merge that with a squash and merge which changed the commit id

That's what I guessed indeed. Would have your release fail on F-Droid (if you want them using your signed APK) as they could not reproduce it.

**Comment by rodvar** [comment]
> > it looks by the way its written that you basically logged your process reviewing
> 
> That's indeed how it was created – with a little cleanup afterwards (e.g. removing most of the errors from the first run, as they were not relevant: the error was mine). I thought doing it this way helps you see how another person would go based on what's there. First-hand experience: if I write up my docs I often take things granted as I know how they work, if I wrote the software. So this could offer a possibility to see it with an "independent pair of eyes" 😉
> 
> > happy to hear the generated apk is fine as is, so we can leave our release process as is with no changes then
> 
> Looks like, yes.
> 
> > the testing apk I provided …
> 
> The process also produced the other APK, but I ignored it for this case. It was more or less a POC, showing the core points (like, the APK should have been built from the tagged commit, the commit needs to be present, it indeed seems to be deterministic, where some instructions would be helpful, the successful recipe to forward).
> 
> > F-droid doesn't support multiple apps per project?
> 
> Which platform does? It would be separate apps. Same on PlayStore, no? If multiple apps are needed, the descriptions need to be clear on that fact. And at least currently, F-Droid is about Android apps, so it would not cover your iosClient at all.
> 
> > and then merge that with a squash and merge which changed the commit id
> 
> That's what I guessed indeed. Would have your release fail on F-Droid (if you want them using your signed APK) as they could not reproduce it.

👍 👍 

> > F-droid doesn't support multiple apps per project?
> 
> Which platform does? It would be separate apps. Same on PlayStore, no? If multiple apps are needed, the descriptions need to be clear on that fact. And at least currently, F-Droid is about Android apps, so it would not cover your iosClient at all.

☝ what I mean is, f-droid will look for fastlane metadata only in root, but for our codebase that doesn't make sense because there are subfolders per app. Any ideas on how can we make that work? Would it be enough to just mention this when I actually do the registration on f-droid?

**Comment by IzzySoft** [comment]
> what I mean is, f-droid will look for fastlane metadata

Ah, **Fastlane** – yes, there are 3 locations they support, as pointed out in the [IzzyOnDroid Fastlane Documentation](https://gitlab.com/IzzyOnDroid/repo/-/wikis/Fastlane):

* `/fastlane/metadata/android/<locale>/` would be the official Fastlane location
* `/metadata/<locale>/` the F-Droid specific location

Both are per repo. The third one is per-flavor: `/src/<buildFlavor>/fastlane/metadata/android/<locale>/`. But that is all.  F-Droid has no setting for "custom locations" with e.g. multiple apps per repo – at least not to my knowledge.

**Comment by endingisnight** [comment]
@IzzySoft I haven't tested this, but looking at [this](https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/update.py#L1036) function, the build metadata's subdir value is used before getting the fastlane/triple-t data. So monorepos with separate apps should work.

**Comment by IzzySoft** [comment]
That's for flavors. Well, go try it. But to my knowledge, the above mentioned 3 locations are the only ones accepted **for fastlane** (Triple-T might be a different horse, I've never really worked with that).

**Comment by rodvar** [comment]
ok great guys, I think next step is to do an app submission to test a distribution in F-Droid. I'll probably start with Bisq Connect app first.

We are still at pre-release stage, I hope that's not an issue. This means this pre-release I'll submit won't be announced or promoted in Bisq official channels yet, until we decide to properly release.

**Comment by rodvar** [comment]
@IzzySoft @endingisnight https://gitlab.com/fdroid/rfp/-/issues/3106 😄

**Comment by endingisnight** [comment]
@rodvar that's great, but you seemingly madr the issue confidencial?

**Comment by rodvar** [comment]
> [@rodvar](https://github.com/rodvar) that's great, but you seemingly madr the issue confidencial?

@endingisnight I thought it was the right thing to do since this is a pre-release specifically to make sure we are able to release on f-droid. Is that ok? Happy to uncheck that if needed

**Comment by rodvar** [comment]
> > I haven't tested this, but looking at [this](https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/update.py#L1036) function, the build metadata's subdir value is used before getting the fastlane/triple-t data. So monorepos with separate apps should work.
> 
> Indeed the subdir is taken into account, however only for Triple-T. Mullvad VPN makes use of it, for instance.
> 
> The [Fastlane extraction function](https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/update.py#L1134) doesn't. I think it probably makes sense for it do so too, though? It would come in handy in this case and I believe [VLC](http://code.videolan.org/videolan/vlc-android/-/issues/1000) might benefit too (I think jspricke somewhat proposed it in #fdroid-dev) and it does seem sensible? I'm very much not an expert though so I might be missing something terribly obvious.

hey @leoheitmannruiz thanks for jumping in. I've posted a comment in the submission request about this, the reviewer suggested to have a look at triple-t and from what I've seen I think we might need to migrate the metadata to that structure, I'm waiting on their confirmation. Thanks

**Comment by IzzySoft** [comment]
@rodvar I'm not involved with the reviews at RFP, so I cannot answer; you'll have to ask that there I guess. Not sure how much Leo is involved with RFP, maybe he can tell.

**Comment by rodvar** [comment]
Suddenly sustained radio silence from F-Droid devs unfortunately.
Moving this one to the backlog for now.

**Comment by rodvar** [comment]
@wodoro this has the history, would you be interested in re-opening and see if we can land it?

There is also [this discussion that happened a bit after](https://github.com/bisq-network/bisq-mobile/issues/1475)

**Comment by wodoro** [comment]
sure, I would like this to reopen and take it
